File: coursera/algorithms-i-princeton/java/src/main/java/Percolation.java
import edu.princeton.cs.algs4.WeightedQuickUnionUF;

import java.util.Arrays;

public class Percolation {
    private int gridColumns, totalCells, virtualTop, virtualBottom;
    private boolean[] openedCells;
    private WeightedQuickUnionUF unionFind;

    public Percolation(int n) {
        gridColumns = n;
        totalCells = n * n;
        virtualTop = totalCells;
        virtualBottom = totalCells + 1;
        openedCells = new boolean[n];
        Arrays.fill(openedCells, false);
        unionFind = new WeightedQuickUnionUF(totalCells + 2);
    }

    public void open(int row, int col) {
        int current = index(row, col);
        int up = current - gridColumns;
        int down = current + gridColumns;
        int left = current - 1;
        int right = current + 1;

        openedCells[current] = true;

        if(up >= 0 && openedCells[up]) {
            unionFind.union(current, up);
        }

        System.out.println(down);
        if((down <= (totalCells - 1)) && openedCells[down]) {
            unionFind.union(current, down);
        }

        if(left >= 0 && line(left) == line(current) && openedCells[left]) {
            unionFind.union(current, left);
        }

        if(line(right) == line(current) && openedCells[right]) {
            unionFind.union(current, right);
        }

        if(line(current) == 0) {
            unionFind.union(current, virtualTop);
        }

        if(line(current) == line(totalCells - 1)) {
            unionFind.union(current, virtualBottom);
        }
    }

    private int index(int row, int col) {
        return (gridColumns * row) + col;
    }

    private int line(int i) {
        return (i / gridColumns);
    }

    public boolean isOpen(int row, int col) {
        return openedCells[index(row, col)];
    }

    public boolean isFull(int row, int col) {
        return unionFind.connected(index(row, col), virtualTop);
    }

    public int numberOfOpenSites() {
        // FIXME: not implemented
        return totalCells - unionFind.count() + 1;
    }

    public boolean percolates() {
        return unionFind.connected(virtualTop, virtualBottom);
    }

    public static void main(String[] args) {
        Percolation p = new Percolation(3);
        p.open(0, 0);
    }
}

File: coursera/algorithms-i-princeton/ruby/percolation.rb
require_relative './quick_find.rb';

class Percolation
  def initialize(grid_columns)
    @grid_columns = grid_columns
    @total_cells = grid_columns * grid_columns
    @virtual_top = @total_cells
    @virtual_bottom = @total_cells + 1
    @opened_cells = Array.new(@total_cells+1, false)
    @opened_cells[@virtual_top] = true
    @opened_cells[@virtual_bottom] = true
    @union_find = QuickFind.new(@total_cells + 2)
  end

  def number_of_open_sites
    @total_cells - @union_find.count + 1
  end

  def open(row, col)
    i = index(row, col)
    @opened_cells[i] = true
    neighbors(i).each do |a|
      @union_find.union(i, a) if @opened_cells[a]
    end
  end

  def neighbors(i)
    result = []

    up = i - @grid_columns
    result << up if up >= 0

    down = i + @grid_columns
    result << down if down <= @total_cells - 1

    left = i - 1
    result << left if line(left) == line(i)

    right = i + 1
    result << right if line(right) == line(i)

    result << @virtual_top if line(i) == 0

    result << @virtual_bottom if line(i) == line(@total_cells - 1)

    result
  end

  def line(i)
    i / @grid_columns
  end

  def index(row, col)
    (@grid_columns * row) + col
  end

  def open?(row, col)
    @opened_cells[index(row, col)]
  end

  def full?(row, col)
    @union_find.connected?(index(row, col), @virtual_top)
  end

  def percolates?
    @union_find.connected?(@virtual_top, @virtual_bottom)
  end
end
